src/main/java/uk/gov/hmcts/reform/security/keyvault/KeyVaultService.java

```diff
@@ -43,49 +30,42 @@
-    KeyVaultService(KeyVaultClient vaultClient, LoadingCache<String, KeyBundle> keyByAliasCache, LoadingCache<String,
-        KeyBundle> keyByIdentifierCache, LoadingCache<String, CertificateBundle> certificateByAliasCache) {
-        baseUrl = System.getProperty(BASE_URL_PROPERTY);
-        this.vaultClient = vaultClient;
-        this.keyByAliasCache = keyByAliasCache;
-        this.keyByIdentifierCache = keyByIdentifierCache;
-        this.certificateByAliasCache = certificateByAliasCache;
+        return INSTANCE;
     }
 
     private KeyVaultService() {
-        baseUrl = System.getProperty(BASE_URL_PROPERTY);
-
-        String clientId = System.getProperty(CLIENT_ID_PROPERTY);
-        String clientSecret = System.getProperty(CLIENT_SECRET_PROPERTY);
+        this(new SystemPropertyKeyVaultConfigBuilder().build());
+    }
 
-        KeyVaultCredentials keyVaultCredentials = new KeyVaultCredentials() {
+    KeyVaultService(KeyVaultConfig keyVaultConfig) {
+        this(keyVaultConfig, new KeyVaultClientProvider() {
             @Override
-            public String doAuthenticate(String authorization, String resource, String scope) {
-                ExecutorService service = null;
-                try {
-                    service = Executors.newFixedThreadPool(1);
-                    AuthenticationContext context = new AuthenticationContext(authorization, false, service);
-
-                    ClientCredential credential = new ClientCredential(clientId, clientSecret);
-                    Future<AuthenticationResult> future = context.acquireToken(resource, credential, null);
-                    AuthenticationResult authenticationResult = future.get(30, TimeUnit.SECONDS);
-
-                    return authenticationResult.getAccessToken();
-                } catch (MalformedURLException | InterruptedException | ExecutionException | TimeoutException e) {
-                    throw new ProviderException(e);
-                } finally {
-                    if (service != null) {
-                        service.shutdown();
-                    }
+            public KeyVaultClient getClient(KeyVaultConfig keyVaultConfig) {
+                if (StringUtils.isNoneEmpty(keyVaultConfig.getVaultClientId(), keyVaultConfig.getVaultClientKey())) {
+                    return new KeyVaultClient(new ClientSecretKeyVaultCredential(keyVaultConfig.getVaultClientId(),
+                        keyVaultConfig.getVaultClientKey()));
+                } else if (StringUtils.isNotEmpty(keyVaultConfig.getVaultMsiUrl())) {
+                    return new KeyVaultClient(new AccessTokenKeyVaultCredential(keyVaultConfig.getVaultMsiUrl(),
+                        keyVaultConfig.getVaultErrorMaxRetries(), keyVaultConfig.getVaultErrorRetryIntervalMillis()));
                 }
+
+                throw new IllegalArgumentException("System properties do not define which KeyVaultClient to create");
             }
-        };
+        }.getClient(keyVaultConfig));
```

Comment:
Creating an anonymous implementation and then immediately calling the getClient method on it feels a bit weird. I wonder if it would be clearer to just have a nested implementation of the provider interface later in the class. Maybe this has all ended up being a bit confusing because the health check code can't create the KeyVaultClient in the constructor, but in this case we can because we already have the properties.